File: backend/core/rerank.py
# ----- equity-aware re-ranking @ backend/core/rerank.py -----
import json
import redis
from typing import Any, Dict, List, Optional, Set
from backend.utils.config import config
from backend.utils.logger import logger

link_graph: Dict[str, int] = {}
LINK_GRAPH_KEY = "autolinks:link_graph"


def init_link_graph(graph: Dict[str, int]) -> None:
    """Initialize the link graph with pre-computed inbound link counts."""
    global link_graph
    link_graph = graph
    _save_link_graph(graph)
    logger.info("Link graph initialized with %d URLs", len(link_graph))


def restore_link_graph() -> Dict[str, int]:
    """Restore the link graph from Redis on startup."""
    global link_graph
    try:
        if not config.redis_url:
            return {}
        rds = _get_redis()
        raw = rds.get(LINK_GRAPH_KEY)
        if raw:
            graph = json.loads(raw)
            link_graph = graph
            logger.info("Link graph restored from Redis: %d URLs", len(graph))
            return graph
    except Exception as e:
        logger.warning("Could not restore link graph from Redis: %s", e)
    return {}


def _save_link_graph(graph: Dict[str, int]) -> None:
    """Persist the link graph to Redis."""
    try:
        if not config.redis_url or not graph:
            return
        rds = _get_redis()
        rds.set(LINK_GRAPH_KEY, json.dumps(graph))
    except Exception as e:
        logger.warning("Could not save link graph to Redis: %s", e)


def _get_redis():
    redis_url = config.redis_url
    kwargs = {}
    if redis_url.startswith("rediss://"):
        kwargs["ssl_cert_reqs"] = None
    return redis.Redis.from_url(redis_url, **kwargs)


def equity_need(inbound_links: int) -> float:
    """
    Calculate equity need score for a URL.

    Args:
        inbound_links: Number of inbound internal links

    Returns:
        Float between 0 and 1, higher = more need
    """
    return 1 / (1 + inbound_links)


def final_score(similarity: float, inbound_links: int, alpha: float = None) -> float:
    """
    Compute final combined score using similarity + equity need.

    Args:
        similarity: Cosine similarity from vector search
        inbound_links: Number of inbound internal links
        alpha: Weight for similarity (1-alpha for equity)

    Returns:
        Combined score
    """
    if alpha is None:
        alpha = config.rerank_alpha

    eq_need = equity_need(inbound_links)
    return alpha * similarity + (1 - alpha) * eq_need


def collapse_candidates_by_url(
    candidates: List[Dict[str, Any]],
) -> List[Dict[str, Any]]:
    """
    Collapse chunk-level search hits into one best candidate per URL.

    Args:
        candidates: List of raw Qdrant chunk hits

    Returns:
        List containing only the highest-scoring chunk for each URL
    """
    best_by_url: Dict[str, Dict[str, Any]] = {}

    for candidate in candidates:
        url = candidate.get("url", "")
        if not url:
            continue

        existing_candidate = best_by_url.get(url)
        if existing_candidate is None or candidate.get(
            "score", 0.0
        ) > existing_candidate.get("score", 0.0):
            best_by_url[url] = candidate

    return list(best_by_url.values())


def rerank_candidates(
    candidates: List[Dict[str, Any]],
    alpha: float = None,
    excluded_urls: Optional[Set[str]] = None,
) -> List[Dict[str, Any]]:
    """
    Re-rank Qdrant results using equity-aware scoring.

    Args:
        candidates: List of {url, chunk_text, score} from Qdrant
        alpha: Similarity weight (default from config)
        excluded_urls: URLs already selected elsewhere in the response

    Returns:
        Re-ranked list with equity_need_score and final_score added
    """
    if alpha is None:
        alpha = config.rerank_alpha

    if excluded_urls is None:
        excluded_urls = set()

    unique_candidates = collapse_candidates_by_url(candidates)
    reranked_candidates = []

    for candidate in unique_candidates:
        url = candidate.get("url", "")
        if url in excluded_urls:
            continue

        inbound_count = link_graph.get(url, 0)
        eq_need = equity_need(inbound_count)
        sim_score = candidate.get("score", 0.0)
        final = final_score(sim_score, inbound_count, alpha)

        reranked_candidates.append(
            {
                **candidate,
                "inbound_link_count": inbound_count,
                "equity_need_score": round(eq_need, 4),
                "final_score": round(final, 4),
            }
        )

    reranked_candidates.sort(
        key=lambda candidate: candidate["final_score"], reverse=True
    )
    logger.info(
        "Re-ranked %s unique URL candidates from %s raw chunks",
        len(reranked_candidates),
        len(candidates),
    )
    return reranked_candidates

File: backend/internal/rerank/rerank.go
// ----- equity-aware re-ranking and link graph @ backend/internal/rerank/rerank.go -----
package rerank

import (
	"context"
	"encoding/json"
	"math"
	"sort"
	"sync"

	"github.com/anomalyco/autolinks/internal/config"
	"github.com/anomalyco/autolinks/internal/logger"
	"github.com/redis/go-redis/v9"
)

// LinkGraphKey is the Redis key for the link graph.
const LinkGraphKey = "autolinks:link_graph"

var (
	linkGraph     map[string]int
	linkGraphMu   sync.RWMutex
	rerankRdb     *redis.Client
	rerankRdbErr  error
	rerankRdbOnce sync.Once
)

func init() {
	linkGraph = make(map[string]int)
}

// InitLinkGraph initializes the link graph with pre-computed inbound link counts.
func InitLinkGraph(graph map[string]int) {
	linkGraphMu.Lock()
	defer linkGraphMu.Unlock()
	linkGraph = graph
	saveLinkGraph(graph)
	logger.Info("Link graph initialized with %d URLs", len(linkGraph))
}

// RestoreLinkGraph restores the link graph from Redis on startup.
func RestoreLinkGraph() map[string]int {
	redisURL := config.RedisURL()
	if redisURL == "" {
		return map[string]int{}
	}

	rdb := getRedisClient()
	ctx := context.Background()
	raw, err := rdb.Get(ctx, LinkGraphKey).Result()
	if err != nil {
		logger.Warning("Could not restore link graph from Redis: %s", err)
		return map[string]int{}
	}

	var graph map[string]int
	if err := json.Unmarshal([]byte(raw), &graph); err != nil {
		logger.Warning("Could not unmarshal link graph from Redis: %s", err)
		return map[string]int{}
	}

	linkGraphMu.Lock()
	linkGraph = graph
	linkGraphMu.Unlock()

	logger.Info("Link graph restored from Redis: %d URLs", len(graph))
	return graph
}

func saveLinkGraph(graph map[string]int) {
	redisURL := config.RedisURL()
	if redisURL == "" || len(graph) == 0 {
		return
	}

	rdb := getRedisClient()
	data, err := json.Marshal(graph)
	if err != nil {
		logger.Warning("Could not save link graph to Redis: %s", err)
		return
	}

	ctx := context.Background()
	if err := rdb.Set(ctx, LinkGraphKey, data, 0).Err(); err != nil {
		logger.Warning("Could not save link graph to Redis: %s", err)
	}
}

func getRedisClient() *redis.Client {
	rerankRdbOnce.Do(func() {
		redisURL := config.RedisURL()
		if redisURL == "" {
			return
		}
		opts, err := redis.ParseURL(redisURL)
		if err != nil {
			logger.Error("Failed to parse Redis URL: %s", err)
			rerankRdbErr = err
			return
		}
		rerankRdb = redis.NewClient(opts)
	})
	return rerankRdb
}

// EquityNeed calculates equity need score for a URL (higher = more need).
func EquityNeed(inboundLinks int) float64 {
	return 1.0 / (1.0 + float64(inboundLinks))
}

// FinalScore computes final combined score using similarity + equity need.
func FinalScore(similarity float64, inboundLinks int, alpha float64) float64 {
	eqNeed := EquityNeed(inboundLinks)
	return alpha*similarity + (1-alpha)*eqNeed
}

// CollapseCandidatesByURL collapses chunk-level search hits into one best candidate per URL.
func CollapseCandidatesByURL(candidates []Candidate) []Candidate {
	bestByURL := make(map[string]Candidate)

	for _, candidate := range candidates {
		url := candidate.URL
		if url == "" {
			continue
		}

		existing, ok := bestByURL[url]
		if !ok || candidate.Score > existing.Score {
			bestByURL[url] = candidate
		}
	}

	var result []Candidate
	for _, c := range bestByURL {
		result = append(result, c)
	}
	return result
}

// Candidate represents a raw Qdrant search hit enriched with equity scores.
type Candidate struct {
	URL              string
	ChunkText        string
	Score            float64
	InboundLinkCount int
	EquityNeedScore  float64
	FinalScore       float64
}

// RerankCandidates re-ranks Qdrant results using equity-aware scoring.
func RerankCandidates(candidates []Candidate, alpha float64, excludedURLs map[string]bool) []Candidate {
	if alpha == 0 {
		alpha = config.RerankAlpha()
	}

	uniqueCandidates := CollapseCandidatesByURL(candidates)

	var reranked []Candidate
	for _, candidate := range uniqueCandidates {
		if excludedURLs != nil && excludedURLs[candidate.URL] {
			continue
		}

		linkGraphMu.RLock()
		inboundCount := linkGraph[candidate.URL]
		linkGraphMu.RUnlock()

		eqNeed := EquityNeed(inboundCount)
		final := FinalScore(candidate.Score, inboundCount, alpha)

		reranked = append(reranked, Candidate{
			URL:              candidate.URL,
			ChunkText:        candidate.ChunkText,
			Score:            candidate.Score,
			InboundLinkCount: inboundCount,
			EquityNeedScore:  math.Round(eqNeed*10000) / 10000,
			FinalScore:       math.Round(final*10000) / 10000,
		})
	}

	sort.Slice(reranked, func(i, j int) bool {
		return reranked[i].FinalScore > reranked[j].FinalScore
	})

	logger.Info("Re-ranked %d unique URL candidates from %d raw chunks", len(reranked), len(candidates))
	return reranked
}

// GetLinkGraph returns a copy of the current link graph.
func GetLinkGraph() map[string]int {
	linkGraphMu.RLock()
	defer linkGraphMu.RUnlock()
	result := make(map[string]int, len(linkGraph))
	for k, v := range linkGraph {
		result[k] = v
	}
	return result
}
